### PR TITLE
Add GitHub Connector

### LINKED ISSUE #35: Github support in unifile
The goal is to offer a new service in unifile, to use github to store your files. The advantages are 
* free (max size of a repository is 1GB, no limit in the number of repos)
* everything is versioned with git
* one can use it to publish on [github pages](https://pages.github.com/) (which is great quality hosting and free by github themselves)

I have tested the github API, which is quite simple to use, without any SDK or library, just the HTTP API, see my little nodejs test:
https://gist.github.com/lexoyo/932550e8ddbae2ec1d2d

But there is the question of the UI and how we handle repos and branches in unifile. First I thought that

* `/xxx` would be the repo xxx
* `/xxx/yyy/` would be the branch yyy of the repo xxx
* `/xxx/yyy/zzz` would be the folder or file zzz on the branch yyy of the repo xxx

But then I thought that the user will try to store files at the repo or branch level, and will not understand what they are doing. So I see several solutions

* add an `info` field in the json response which tells the user that "at this level, you can not store files, you need to have only folders" or do not even allow to create folders for repos and branches but say "you can not create files or folders here, you need to create a repository in your github acount"
* when the user has authorized unifile to access their github account, ask him to choose a repo (a bit like in open pages when we ask for a name)

The next question would then be how to make people understand how to publish on github pages, they need to publish their site on a branch called gh-pages...

### PR DESCRIPTION
**Pull request by JbIPS**
Add a connector to GitHub via their HTTP API

Treat repos and branches as parent directories. Syntax looks like /repo/branch/parentFolder/folder
With an empty path it list all repos. A path of one level list all branches in one repo. Further path list files and folder in this repo, on this branch and at this exact path.

This add a constraint on `PUT` command because we can only create a file on the 3rd level. If less are provided, it will return an error.

Close #35

### CODE REVIEW

**Comment by JbIPS** [comment]
Wait before merge, I'll just change my config with the newly merged update and add GitHub to the README

**Comment by JbIPS** [comment]
Ok, now you can merge!

**Comment by lexoyo** [comment]
Great work @JbIPS !

Little bugs I see
* no more than 30 repositories are displayed (just display them all if posible?)
* not possible to create a 1st branch after creating a repo 
* in cloud explorer, there is no way to upload a file, maybe because the file name is in POST (write content from the URL works)
* write content to a branch other than master fails (returns success though)
* get image or open an HTML page in Silex fails, because the returned data is json instead of the file content
* remove a folder does nothing but returns true

File: lib/services/github.js [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+ * }
+ */
+exports.get = function (path, request, response, next, cbk) {
+	if (!isConnected(request)){
+		cbk({success:false, code:401, message:'User not logged in yet. You need to call the \'login\' service first.'});
+	}
+	else{
+		var splitPath = getPathTokens(path);
+		if (splitPath.length < 3) {
+			return cbk({success: false, message: 'Invalid path. It must at least have 3 levels.'});
```
Comment (lexoyo) [inline lib/services/github.js]: another proposal for this text:

This folder can only contain folders not files. Files can go in sub-folders. 
2 levels are required to store files on github: repository > branch > actual folders and files.

Comment (JbIPS) [inline lib/services/github.js]: Isn't that a bit long? Maybe just `This folder can only contain folders not files. Files can go in sub-folders. `

**Comment by JbIPS** [comment]
Thank you for the tests, you found lots of cases I didn't try! I'll fix this ASAP

**Comment by JbIPS** [comment]
@lexoyo About 
> not possible to create a 1st branch after creating a repo 
it's because we can't create refs on an empty repo. My solution is to init the repo on creation with `auto_init` option (create a readme on master) then remove the file but leave the branch.
This also means we have to deny a branch suppression if it's the last one (which would bring us in the same state than a new repo).

What do you think?

**Comment by lexoyo** [comment]
>> not possible to create a 1st branch after creating a repo

> it's because we can't create refs on an empty repo. My solution is to init the repo on creation with auto_init option (create a readme on master) then remove the file but leave the branch.

Ok I see now :)
Your solution is good. You could even leave the readme, and maybe put some useful info in it...

> This also means we have to deny a branch suppression if it's the last one (which would bring us in the same state than a new repo).

Yes ineed, let's prevent it with an error message like "you can not leave this folder empty, it has to have at least 1 folder in it" ?

**Comment by JbIPS** [comment]
I fixed all your bugs but I still have troubles uploading an image. Did you had this problem too? Do you know any workaround?

**Comment by lexoyo** [comment]
No I don't think so. Did you check the corresponding code in the other
services?

**Comment by JbIPS** [comment]
I did it! I had to manually create a base64 encoded blob and that do the trick!

But another problem I spotted is the batch operation with CE: it's calling my functions in parallel instead of with an array which could allow me to create only one commit for all the changes. What do you think of that?

**Comment by lexoyo** [comment]
> I did it! I had to manually create a base64 encoded blob and that do the trick!

Ok well done! Is it github API which is expecting base64 encoded blob?

> But another problem I spotted is the batch operation with CE: it's calling my functions in parallel instead of with an array which could allow me to create only one commit for all the changes. What do you think of that?

You are talking about the multi-file upload?

If you wait 500ms before actually do a commit and buffer the upload/save operations maybe... Would that help?

**Comment by JbIPS** [comment]
Blobs can be utf8 (default) or base64 encoded. But the utf8 encoded version of an image is badly decoded (don't know why).So forcing to base64 did the trick.

> You are talking about the multi-file upload?

I'm taking all batch operation, like deleting or uploading. I could use a buffer strategy but this will slow down those operation for nothing in the majority of the case.

**Comment by lexoyo** [comment]
I can not think to a simple solution... We should change CE and unifile to enable batch operations, and commit message too.

Also when using it with silex, this gives commits like this one
https://github.com/singchan/silex-templates/commit/d71300734cb0310c0d46f1a830c240cc4299a57a

A few other minor bugs, I'll create issues in a while

The only important bug I see is the one where you have to reload the html page to display your files in CE.

**Comment by lexoyo** [comment]
Did you have a chance to try?

**Pull request merged**